In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns
import statsmodels.api as sm
from scipy.stats import boxcox
from itertools import permutations,combinations
from collections import Counter
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"
pd.options.display.max_rows = 100

/opt/anaconda3/envs/DA_Env/lib/python3.9/site-packages/statsmodels/compat/pandas.py:61: FutureWarning: pandas.Int64Index is deprecated and will be removed from pandas in a future version. Use pandas.Index with the appropriate dtype instead.
  from pandas import Int64Index as NumericIndex


In [3]:
articles = pd.read_csv ('articles.csv')
articles

article_id  product_code               prod_name  product_type_no  \
0        108775015        108775               Strap top              253   
1        108775044        108775               Strap top              253   
2        108775051        108775           Strap top (1)              253   
3        110065001        110065       OP T-shirt (Idro)              306   
4        110065002        110065       OP T-shirt (Idro)              306   
...            ...           ...                     ...              ...   
105537   953450001        953450  5pk regular Placement1              302   
105538   953763001        953763       SPORT Malaga tank              253   
105539   956217002        956217         Cartwheel dress              265   
105540   957375001        957375        CLAIRE HAIR CLAW               72   
105541   959461001        959461            Lounge dress              265   

       product_type_name  product_group_name  graphical_appearance_no  \
0               Vest top  Garment Upper body                  1010016   
1               Vest top  Garment Upper body                  1010016   
2               Vest top  Garment Upper body                  1010017   
3                    Bra           Underwear                  1010016   
4                    Bra           Underwear                  1010016   
...                  ...                 ...                      ...   
105537             Socks      Socks & Tights                  1010014   
105538          Vest top  Garment Upper body                  1010016   
105539             Dress   Garment Full body                  1010016   
105540         Hair clip         Accessories                  1010016   
105541             Dress   Garment Full body                  1010016   

       graphical_appearance_name  colour_group_code colour_group_name  ...  \
0                          Solid                  9             Black  ...   
1                          Solid                 10             White  ...   
2                         Stripe                 11         Off White  ...   
3                          Solid                  9             Black  ...   
4                          Solid                 10             White  ...   
...                          ...                ...               ...  ...   
105537           Placement print                  9             Black  ...   
105538                     Solid                  9             Black  ...   
105539                     Solid                  9             Black  ...   
105540                     Solid                  9             Black  ...   
105541                     Solid                 11         Off White  ...   

          department_name index_code        index_name index_group_no  \
0            Jersey Basic          A        Ladieswear              1   
1            Jersey Basic          A        Ladieswear              1   
2            Jersey Basic          A        Ladieswear              1   
3          Clean Lingerie          B  Lingeries/Tights              1   
4          Clean Lingerie          B  Lingeries/Tights              1   
...                   ...        ...               ...            ...   
105537          Socks Bin          F          Menswear              3   
105538             Jersey          A        Ladieswear              1   
105539             Jersey          A        Ladieswear              1   
105540  Small Accessories          D           Divided              2   
105541             Jersey          A        Ladieswear              1   

        index_group_name section_no            section_name garment_group_no  \
0             Ladieswear         16  Womens Everyday Basics             1002   
1             Ladieswear         16  Womens Everyday Basics             1002   
2             Ladieswear         16  Womens Everyday Basics             1002   
3             Ladieswear         61         Womens Lingerie             

Searching which column has NaN

In [4]:
articles.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 105542 entries, 0 to 105541
Data columns (total 25 columns):
 #   Column                        Non-Null Count   Dtype 
---  ------                        --------------   ----- 
 0   article_id                    105542 non-null  int64 
 1   product_code                  105542 non-null  int64 
 2   prod_name                     105542 non-null  object
 3   product_type_no               105542 non-null  int64 
 4   product_type_name             105542 non-null  object
 5   product_group_name            105542 non-null  object
 6   graphical_appearance_no       105542 non-null  int64 
 7   graphical_appearance_name     105542 non-null  object
 8   colour_group_code             105542 non-null  int64 
 9   colour_group_name             105542 non-null  object
 10  perceived_colour_value_id     105542 non-null  int64 
 11  perceived_colour_value_name   105542 non-null  object
 12  perceived_colour_master_id    105542 non-null  int64 
 13 

In [5]:
articles[articles['detail_desc'].isna()==True]

article_id  product_code                   prod_name  product_type_no  \
1467    351332007        351332        Marshall Lace up Top              252   
2644    420049002        420049                  OL TAGE PQ               87   
2645    420049003        420049                  OL TAGE PQ               87   
2742    426199002        426199  Ellen Shortie Daisy Low 3p              286   
2743    426199010        426199  Ellen Shortie Daisy Low 3p              286   
...           ...           ...                         ...              ...   
67838   752458001        752458           Poissy boho dress              265   
72720   768842001        768842                 Andrews set              270   
72721   768842004        768842                 Andrews set              270   
93144   856985001        856985                   Pogo rope               67   
95624   867200001        867200              Necklace Simon               77   

      product_type_name  product_group_name  graphical_appearance_no  \
1467            Sweater  Garment Upper body                  1010018   
2644              Boots               Shoes                  1010016   
2645              Boots               Shoes                  1010016   
2742   Underwear bottom           Underwear                  1010016   
2743   Underwear bottom           Underwear                  1010017   
...                 ...                 ...                      ...   
67838             Dress   Garment Full body                  1010007   
72720       Garment Set   Garment Full body                  1010017   
72721       Garment Set   Garment Full body                  1010004   
93144              Belt         Accessories                  1010016   
95624          Necklace         Accessories                  1010011   

      graphical_appearance_name  colour_group_code colour_group_name  ...  \
1467                  Treatment                  7              Grey  ...   
2644                      Solid                 13             Beige  ...   
2645                      Solid                 23       Dark Yellow  ...   
2742                      Solid                  9             Black  ...   
2743                     Stripe                  8         Dark Grey  ...   
...                         ...                ...               ...  ...   
67838                Embroidery                 10             White  ...   
72720                    Stripe                 10             White  ...   
72721                     Check                 73         Dark Blue  ...   
93144                     Solid                 12       Light Beige  ...   
95624                  Metallic                  3            Silver  ...   

         department_name index_code          index_name index_group_no  \
1467     Jersey Fancy DS          D             Divided              2   
2644     Premium Quality          C  Ladies Accessories              1   
2645     Premium Quality          C  Ladies Accessories              1   
2742     Casual Lingerie          B    Lingeries/Tights              1   
2743     Casual Lingerie          B    Lingeries/Tights              1   
...                  ...        ...                 ...            ...   
67838              Dress          A          Ladieswear              1   
72720     Baby Boy Woven          G    Baby Sizes 50-98              4   
72721     Baby Boy Woven          G    Baby Sizes 50-98              4   
93144              Belts          C  Ladies Accessories              1   
95624  Small Accessories          F            Menswear              3   

       index_group_name section_no            section_name garment_group_no  \
1467            Divided         58        Divided Selected             1005   
2644         Ladieswear         64            Womens Shoes             1020   
2645         Ladieswear         64            Womens Shoes             1020   
2742         Ladieswear         61         Womens L

**Total NaN in each column.**
As we can see only one column has missing values and the nº of missing values is no significant as only 416 from
105542 values are missing.

In [6]:
articles.isna().sum()

article_id                        0
product_code                      0
prod_name                         0
product_type_no                   0
product_type_name                 0
product_group_name                0
graphical_appearance_no           0
graphical_appearance_name         0
colour_group_code                 0
colour_group_name                 0
perceived_colour_value_id         0
perceived_colour_value_name       0
perceived_colour_master_id        0
perceived_colour_master_name      0
department_no                     0
department_name                   0
index_code                        0
index_name                        0
index_group_no                    0
index_group_name                  0
section_no                        0
section_name                      0
garment_group_no                  0
garment_group_name                0
detail_desc                     416
dtype: int64

In [7]:
customers = pd.read_csv ('customers.csv')
customers

customer_id   FN  Active  \
0        00000dbacae5abe5e23885899a1fa44253a17956c6d1c3...  NaN     NaN   
1        0000423b00ade91418cceaf3b26c6af3dd342b51fd051e...  NaN     NaN   
2        000058a12d5b43e67d225668fa1f8d618c13dc232df0ca...  NaN     NaN   
3        00005ca1c9ed5f5146b52ac8639a40ca9d57aeff4d1bd2...  NaN     NaN   
4        00006413d8573cd20ed7128e53b7b13819fe5cfc2d801f...  1.0     1.0   
...                                                    ...  ...     ...   
1371975  ffffbbf78b6eaac697a8a5dfbfd2bfa8113ee5b403e474...  NaN     NaN   
1371976  ffffcd5046a6143d29a04fb8c424ce494a76e5cdf4fab5...  NaN     NaN   
1371977  ffffcf35913a0bee60e8741cb2b4e78b8a98ee5ff2e6a1...  1.0     1.0   
1371978  ffffd7744cebcf3aca44ae7049d2a94b87074c3d4ffe38...  1.0     1.0   
1371979  ffffd9ac14e89946416d80e791d064701994755c3ab686...  NaN     NaN   

        club_member_status fashion_news_frequency   age  \
0                   ACTIVE                   NONE  49.0   
1                   ACTIVE                   NONE  25.0   
2                   ACTIVE                   NONE  24.0   
3                   ACTIVE                   NONE  54.0   
4                   ACTIVE              Regularly  52.0   
...                    ...                    ...   ...   
1371975             ACTIVE                   NONE  24.0   
1371976             ACTIVE                   NONE  21.0   
1371977             ACTIVE              Regularly  21.0   
1371978             ACTIVE              Regularly  18.0   
1371979         PRE-CREATE                   NONE  65.0   

                                               postal_code  
0        52043ee2162cf5aa7ee79974281641c6f11a68d276429a...  
1        2973abc54daa8a5f8ccfe9362140c63247c5eee03f1d93...  
2        64f17e6a330a85798e4998f62d0930d14db8db1c054af6...  
3        5d36574f52495e81f019b680c843c443bd343d5ca5b1c2...  
4        25fa5ddee9aac01b35208d01736e57942317d756b32ddd...  
...                                                    ...  
1371975  7aa399f7e669990daba2d92c577b52237380662f36480b...  
1371976  3f47f1279beb72215f4de557d950e0bfa73789d24acb5e...  
1371977  4563fc79215672cd6a863f2b4bf56b8f898f2d96ed590e...  
1371978  8892c18e9bc3dca6aa4000cb8094fc4b51ee8db2ed14d7...  
1371979  0a1a03306fb2f62164c2a439b38c0caa64b40deaae8687...  

[1371980 rows x 7 columns]

In [8]:
customers.isna().sum()

customer_id                    0
FN                        895050
Active                    907576
club_member_status          6062
fashion_news_frequency     16009
age                        15861
postal_code                    0
dtype: int64

In [9]:
customers_backup = customers.copy()
customers[['FN','Active']] = customers[['FN','Active']].fillna(0)
customers

customer_id   FN  Active  \
0        00000dbacae5abe5e23885899a1fa44253a17956c6d1c3...  0.0     0.0   
1        0000423b00ade91418cceaf3b26c6af3dd342b51fd051e...  0.0     0.0   
2        000058a12d5b43e67d225668fa1f8d618c13dc232df0ca...  0.0     0.0   
3        00005ca1c9ed5f5146b52ac8639a40ca9d57aeff4d1bd2...  0.0     0.0   
4        00006413d8573cd20ed7128e53b7b13819fe5cfc2d801f...  1.0     1.0   
...                                                    ...  ...     ...   
1371975  ffffbbf78b6eaac697a8a5dfbfd2bfa8113ee5b403e474...  0.0     0.0   
1371976  ffffcd5046a6143d29a04fb8c424ce494a76e5cdf4fab5...  0.0     0.0   
1371977  ffffcf35913a0bee60e8741cb2b4e78b8a98ee5ff2e6a1...  1.0     1.0   
1371978  ffffd7744cebcf3aca44ae7049d2a94b87074c3d4ffe38...  1.0     1.0   
1371979  ffffd9ac14e89946416d80e791d064701994755c3ab686...  0.0     0.0   

        club_member_status fashion_news_frequency   age  \
0                   ACTIVE                   NONE  49.0   
1                   ACTIVE                   NONE  25.0   
2                   ACTIVE                   NONE  24.0   
3                   ACTIVE                   NONE  54.0   
4                   ACTIVE              Regularly  52.0   
...                    ...                    ...   ...   
1371975             ACTIVE                   NONE  24.0   
1371976             ACTIVE                   NONE  21.0   
1371977             ACTIVE              Regularly  21.0   
1371978             ACTIVE              Regularly  18.0   
1371979         PRE-CREATE                   NONE  65.0   

                                               postal_code  
0        52043ee2162cf5aa7ee79974281641c6f11a68d276429a...  
1        2973abc54daa8a5f8ccfe9362140c63247c5eee03f1d93...  
2        64f17e6a330a85798e4998f62d0930d14db8db1c054af6...  
3        5d36574f52495e81f019b680c843c443bd343d5ca5b1c2...  
4        25fa5ddee9aac01b35208d01736e57942317d756b32ddd...  
...                                                    ...  
1371975  7aa399f7e669990daba2d92c577b52237380662f36480b...  
1371976  3f47f1279beb72215f4de557d950e0bfa73789d24acb5e...  
1371977  4563fc79215672cd6a863f2b4bf56b8f898f2d96ed590e...  
1371978  8892c18e9bc3dca6aa4000cb8094fc4b51ee8db2ed14d7...  
1371979  0a1a03306fb2f62164c2a439b38c0caa64b40deaae8687...  

[1371980 rows x 7 columns]

In [3]:
transactions = pd.read_csv('transactions_train.csv')
transactions

NameError: name 'pd' is not defined

In [13]:
transactions.isna().sum()

t_dat               0
customer_id         0
article_id          0
price               0
sales_channel_id    0
dtype: int64

In [1]:
basket = transactions.groupby("customer_id").agg({'article_id':'count','price': 'sum'}).reset_index()
basket.columns = ["customer_id", "units", "order_price"]

NameError: name 'transactions' is not defined

NameError: name 'transactions' is not defined